arangod/Aql/grammar.y

```diff
@@ -1703,14 +1683,54 @@ view_name:
   | T_QUOTED_STRING {
       $$ = parser->ast()->createNodeView($1.value);
     }
-  ; // TODO: add parameter case
+  | bind_view {
+      $$ = $1;
+    }
+  ;
 
 bind_parameter:
-    T_PARAMETER {
+    T_DATA_SOURCE_PARAMETER {
+      if ($1.length < 2 || $1.value[0] != '@') {
+        parser->registerParseError(TRI_ERROR_QUERY_BIND_PARAMETER_TYPE, TRI_errno_string(TRI_ERROR_QUERY_BIND_PARAMETER_TYPE), $1.value, yylloc.first_line, yylloc.first_column);
+      }
+
+      auto const param = AstNode::encodeDataSourceType(
+        $1.value, $1.length, AstNode::DataSourceType::Collection
+      );
+
+      auto const* internedParam = parser->query()->registerString(param);
+
+      if (!internedParam) {
+        ABORT_OOM
+      }
+
+      $$ = parser->ast()->createNodeParameter(internedParam, param.size());
```

Comment:
Wouldn't it make sense to add a new type "NODE_TYPE_DATA_SOURCE_PARAMETER" for this?
This would allow us to eliminate all the string suffix magic that is currently there only to make this work with bind parameters of arbitrary types (values, collection names, view names).

Reply:
That was one of the possible approaches but it requires to add extra data member to `AstNode` (which is not desirable I think) or stick with the magic suffixes. As was agreed for the time being we use suffixes and after introducing common parent for `LogicalCollection` and `LogicalView` we move logic to `Ast::injectBindParameters`

Reply:
I was more thinking about adding another node type, e.g. NODE_TYPE_DATA_SOURCE_PARAMETER. It will not require adding an extra data member to AstNode.

Reply:
But how do you want to differ collection from the view then?

Reply:
First a new type NODE_TYPE_DATA_SOURCE_PARAMETER will allow us to tell value bind parameters apart from view/collection name bind parameters.
And then, as each AstNode already has some optional value attached, we could easily give it a type (e.g. by using `AstNode::setIntValue` with value 0 for a collection and value 1 for a view).

Reply:
Ok, but where do you plan to store collection/view name then?

Reply:
Or you want to store the type of the data store in `AstNode::computeValue`?

Reply:
Ah ok, you are right. If we use the string component of value for the view/collection name we cannot use the int component too.
Can't we then convert the parameter to a specialized type once we found out if it's view- or collection-related, like NODE_TYPE_PARAMETER_VIEW and NODE_TYPE_PARAMETER_COLLECTION?

Reply:
Since view is pluggable we eventually should have a node type per view (e.g. NODE_TYPE_PARAMETER_ARANGO_SEARCH_VIEW, NODE_TYPE_PARAMETER_GRAPH_VIEW, etc) which means that we have to patch grammar every time we adding a new view object.
That's why I think it's better to store this information somewhere in `AstNode`.
Or, which is better, we can use new base class we discussed `LogicalDataSource`, but as agreed, it requires a lot more work.